File: src/main/java/eu/luminis/sample/OrderAggregate.java
package eu.luminis.sample;

import eu.luminis.sample.commands.ChangeShippingAddressCommand;
import eu.luminis.sample.commands.CreateOrderCommand;
import eu.luminis.sample.commands.ShipOrderCommand;
import eu.luminis.sample.events.OrderCreatedEvent;
import eu.luminis.sample.events.OrderEvent;
import eu.luminis.sample.events.OrderShippedEvent;
import eu.luminis.sample.events.ShippingAddressChangedEvent;
import eu.luminis.sample.exceptions.AlreadyShippedException;

import java.time.Instant;
import java.util.List;

public class OrderAggregate {

    private boolean hasBeenShipped;

    public static OrderCreatedEvent handle(CreateOrderCommand command) {
        return new OrderCreatedEvent(Instant.now(), command.items(), command.shippingAddress());
    }

    public OrderAggregate(List<OrderEvent> history) {
        if (!firstEventIsOrderCreated(history)) {
            throw new IllegalStateException("Missing order created event");
        }

        applyStateFrom(history);
    }

    public ShippingAddressChangedEvent handle(ChangeShippingAddressCommand command) {
        if (hasBeenShipped) throw new AlreadyShippedException();

        return new ShippingAddressChangedEvent(command.shippingAddress());
    }

    public OrderShippedEvent handle(ShipOrderCommand command) {
        if (hasBeenShipped) throw new AlreadyShippedException();

        return new OrderShippedEvent(Instant.now(), command.shippingProvider());
    }

    private void applyStateFrom(List<OrderEvent> events) {
        for (OrderEvent e : events) {
            switch (e) {
                case OrderCreatedEvent event -> {
                }
                case ShippingAddressChangedEvent event -> {
                }
                case OrderShippedEvent event -> hasBeenShipped = true;
            }
        }
    }

    private static boolean firstEventIsOrderCreated(List<OrderEvent> history) {
        return !history.isEmpty() && history.get(0) instanceof OrderCreatedEvent;
    }
}

File: Project/OrderAggregate.cs
using OrderAggregate.Command;
using OrderAggregate.Events;
using OrderAggregate.Exceptions;

namespace OrderAggregate;

public class OrderAggregate
{
    private bool _hasBeenShipped;

    public static OrderCreatedEvent Handle(CreateOrderCommand command)
    {
        return new OrderCreatedEvent(DateTime.Now, command.Items, command.ShippingAddress);
    }

    public OrderAggregate(List<IOrderEvent> history)
    {
        if (!FirstEventIsOrderCreated(history))
        {
            throw new InvalidOperationException("Missing order created event");
        }

        ApplyStateFrom(history);
    }

    public ShippingAddressChangedEvent Handle(ChangeShippingAddressCommand command)
    {
        if (_hasBeenShipped) throw new AlreadyShippedException();

        return new ShippingAddressChangedEvent(command.ShippingAddress);
    }

    public OrderShippedEvent Handle(ShipOrderCommand command)
    {
        if (_hasBeenShipped) throw new AlreadyShippedException();

        return new OrderShippedEvent(DateTime.Now, command.ShippingProvider);
    }

    private void ApplyStateFrom(List<IOrderEvent> events)
    {
        foreach (var e in events)
        {
            switch (e) 
            {
                case OrderCreatedEvent:
                {
                    break;
                }
                case ShippingAddressChangedEvent:
                {
                    break;
                }
                case OrderShippedEvent:
                {
                    _hasBeenShipped = true;
                    break;
                }
            }
        }
    }

    private static bool FirstEventIsOrderCreated(List<IOrderEvent> history)
    {
        return history.Any() && history[0] is OrderCreatedEvent;
    }
}
